File: practice/convert/example2_book_library/java/book-library/src/main/java/dev/starryeye/book_library/domain/user/loan_history/UserLoanHistory.java
package dev.starryeye.book_library.domain.user.loan_history;

import dev.starryeye.book_library.domain.BaseEntity;
import dev.starryeye.book_library.domain.user.User;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

@Getter
@Entity
@Table(name = "user_loan_history")
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@EntityListeners(value = AuditingEntityListener.class)
public class UserLoanHistory extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    private Long bookId;

    private Boolean isReturn;

    @Builder
    private UserLoanHistory(Long id, User user, Long bookId, Boolean isReturn) {
        if (user == null) {
            throw new IllegalArgumentException("user must not be null");
        }
        if (bookId == null) {
            throw new IllegalArgumentException("bookId must not be null");
        }
        if (isReturn == null) {
            throw new IllegalArgumentException("loanedAt must not be null");
        }

        this.id = id;
        this.user = user;
        this.bookId = bookId;
        this.isReturn = isReturn;
    }

    public static UserLoanHistory create(User user, Long bookId) {
        return UserLoanHistory.builder()
                .id(null)
                .user(user)
                .bookId(bookId)
                .isReturn(false)
                .build();
    }

    public void markReturned() {
        if (isReturn == null) {
            throw new IllegalStateException("history status is wrong.., id = " + id);
        }
        this.isReturn = true;
    }

    public boolean isNotReturned() {
        return !this.isReturn;
    }
}

File: practice/convert/example2_book_library/kotlin/book-library/src/main/kotlin/dev/starryeye/book_library/domain/user/loan_history/UserLoanHistory.kt
package dev.starryeye.book_library.domain.user.loan_history

import dev.starryeye.book_library.domain.BaseEntity
import dev.starryeye.book_library.domain.user.User
import jakarta.persistence.*
import org.springframework.data.jpa.domain.support.AuditingEntityListener

@Entity
@Table(name = "user_loan_history")
@EntityListeners(AuditingEntityListener::class)
class UserLoanHistory(
    @field:ManyToOne(fetch = FetchType.LAZY, optional = false)
    @field:JoinColumn(name = "user_id", nullable = false)
    val user: User,

    val bookId: Long,

    @field:Enumerated(EnumType.STRING)
    var status: UserLoanStatus = UserLoanStatus.LOANED,
) : BaseEntity() {

    @field:Id
    @field:GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null

    fun markReturned() {
        if (status == UserLoanStatus.RETURNED) {
            throw IllegalStateException("already returned, id=$id")
        }
        this.status = UserLoanStatus.RETURNED
    }

    fun isNotReturned(): Boolean = status == UserLoanStatus.LOANED

    companion object {
        fun fixture(
            user: User = User.fixture(),
            bookId: Long = 1L,
            status: UserLoanStatus = UserLoanStatus.LOANED,
        ): UserLoanHistory {
            return UserLoanHistory(
                user = user,
                bookId = bookId,
                status = status,
            )
        }
    }
}
